### ISSUE TITLE
Support passing symbols to whenInjectedInto

### ISSUE THREAD

**Issue by remojansen** [inversify/InversifyJS#599]
`whenInjectedInto` should support service identifiers:
- string
- symbol (needs to be implemented)
- class 

Reported by @akosyakov.

**Comment by theodesp** [comment]
That looks interesting. If I'm not overloaded with work this week I might work on it. What insights can you give for this?

**Comment by remojansen** [comment]
We actually need to change thre APIs:

- [`whenInjectedInto`](https://github.com/inversify/InversifyJS/blob/master/src/syntax/binding_when_syntax.ts#L42)
- [`whenAnyAncestorIs`](https://github.com/inversify/InversifyJS/blob/master/src/syntax/binding_when_syntax.ts#L63)
- [`whenNoAncestorIs`](https://github.com/inversify/InversifyJS/blob/master/src/syntax/binding_when_syntax.ts#L70)

Instead of taking `(Function|string)` they should allow [`ServiceIdentifier<T>`](https://github.com/inversify/InversifyJS/blob/master/src/interfaces/interfaces.ts#L40).

```ts
type ServiceIdentifier<T> = (string | symbol | Newable<T> | Abstract<T>);
```
This means that **4 different types are allowed as arguments**.

So the APIs will become:

- `whenInjectedInto(parent: ServiceIdentifier<T>) => interfaces.BindingOnSyntax<T>`
- `whenAnyAncestorIs(parent: ServiceIdentifier<T>) => interfaces.BindingOnSyntax<T>`
- `whenNoAncestorIs(parent: ServiceIdentifier<T>) => interfaces.BindingOnSyntax<T>`

All these functions use [`typeConstraint`](https://github.com/inversify/InversifyJS/blob/master/src/syntax/constraint_helpers.ts#L33-L51). The `typeConstraint` function can handle the types `string` and `Newable<T>` cases. 

- We need unit tests to make sure that it also works in the `Abstract<T>` case (it should be [covered](https://github.com/inversify/InversifyJS/blob/master/src/syntax/constraint_helpers.ts#L44-L47)).  

- We also need to add support for the `symbol` case which means that we need a new [`} else if() {`](https://github.com/inversify/InversifyJS/blob/master/src/syntax/constraint_helpers.ts#L44).

**Comment by theodesp** [comment]
So I should be able to add additional check to check for Symbol Equality by value possibly using `Symbol.keyFor` or `Symbol.for` ( I need to check for browser support) and perform the check and as for the `Abstract<T>` I should be able to compare their `.prototype` property.

I'm looking at the test cases here
https://github.com/inversify/InversifyJS/blob/1bdb3ac5b4ac7c756cbf90998a0492c155aa0765/test/inversify.test.ts#L2264

and I should be able to add more tests based on this. 

Is that a valid plan?

**Comment by remojansen** [comment]
Yes, the plan sounds good for me :+1: For the test cases create a new file under the [features](https://github.com/inversify/InversifyJS/tree/master/test/features) folder (the `inversify.test.ts` is too long...).

**Comment by theodesp** [comment]
I would like to ask what are the plans for the binding type.
For example I have the following scenario:
```
 let TYPES = {
        Ninja: "Ninja",
        NinjaMaster: Symbol("NinjaMaster"),
        NinjaStudent: Symbol("NinjaStudent"),
        Weapon: "Weapon",
    };

container.bind<Ninja>(TYPES.Ninja).to(NinjaMaster).whenTargetNamed("lethal");
container.bind<Ninja>(TYPES.Ninja).to(NinjaStudent).whenTargetNamed("non-lethal");
container.bind<Weapon>(TYPES.Weapon).to(Katana).whenInjectedInto(TYPES.NinjaMaster);
container.bind<Weapon>(TYPES.Weapon).to(Bokken).whenInjectedInto(TYPES.NinjaStudent);

let master = container.getNamed<Ninja>(TYPES.Ninja, "lethal");
let student = container.getNamed<Ninja>(TYPES.Ninja, "non-lethal");
```
That will fail because
```
Error: No matching bindings found for serviceIdentifier: Weapon
Registered bindings:
 Katana
 Bokken
```

When the compiler assembles the plan it will go over the bindings and it Will try to do a Request using  TYPES.Ninja as a service identifier on the left-hand side and calling
```
typeConstraint(parent)(request.parentRequest);
```
when it comes to narrow the bindings.
In the first case **parent** is **TYPES.NinjaMaster** and **request.parentRequest** is the **NinjaMaster**

So it will try to compare TYPES.NinjaMaster with TYPES.Ninja to see if there is a match. How can I make sure that I Find the correct binding for the correct target?

As far as I understand we need to Match `.to(NinjaMaster)` with the constraint `whenInjectedInto(TYPES.NinjaMaster)`

**Comment by remojansen** [comment]
Hi @theodesp  sorry for the late reply on this. I will get back to you ASAP (probably this weekend)

**Comment by theodesp** [comment]
Don't worry Remo. I'm looking forward to your new book! I have a Safari books online Subscription!

**Comment by theodesp** [comment]
Any further reply on this? hacktoberfest is live!

**Comment by remojansen** [comment]
Hi @theodesp sorry for the late reply, the book is taking all my time :cry: 

In your example, you could solve the problem by using `whenParentNamed` instead of `whenInjectedInto`:

```ts
 let TYPES = {
    Ninja: "Ninja",
    NinjaMaster: Symbol("NinjaMaster"),
    NinjaStudent: Symbol("NinjaStudent"),
    Weapon: "Weapon",
};

let TAGS = {
    lethal: "lethal",
    nonLethal: "non-lethal"
}

container.bind<Ninja>(TYPES.Ninja).to(NinjaMaster).whenTargetNamed(TAGS.lethal);
container.bind<Ninja>(TYPES.Ninja).to(NinjaStudent).whenTargetNamed(TAGS.nonLethal);
container.bind<Weapon>(TYPES.Weapon).to(Katana).whenParentNamed(TAGS.lethal); // here!
container.bind<Weapon>(TYPES.Weapon).to(Bokken).whenParentNamed(TAGS.nonLethal); // here!

let master = container.getNamed<Ninja>(TYPES.Ninja, "lethal");
let student = container.getNamed<Ninja>(TYPES.Ninja, "non-lethal");
```
If the `whenInjectedInto` constraint is checking the type identifier maybe it could be solved by using the class constructor (it can be accessed via the requests.binding.constructor property) using `whenInjectedInto(NinjaMaster)` and `whenInjectedInto(NinjaStudent)`.

Thanks for helping out :+1:
